### ISSUE TITLE
LAA playback for later LucasArts games sounds way off

### ISSUE THREAD

**Issue by JPLeBreton** [adplug/adplug#12]
Resurrecting this old bug from the SF site:

https://sourceforge.net/p/adplug/bugs/23/

The music sample links provided are long dead, I've attached the LAA file which you can compare with the music in this:
https://youtu.be/2BjpCc9Wcb8?t=37s

[mi2intro.zip](https://github.com/adplug/adplug/files/170606/mi2intro.zip)

No OPL player program I know of can play back these files correctly, it'd be nice if one could.

**Comment by Malvineous** [comment]
Can you supply a patch to correct the issue?  I'm happy to merge any fixes but I don't have any experience myself with this format.

**Comment by JPLeBreton** [comment]
I don't have any knowledge of the file format or OPL programming, so sorry I have no patch right now.

Looks like ScummVM forum user "tashiy" was looking into this a while ago:
http://forums.scummvm.org/viewtopic.php?t=11625&postdays=0&postorder=asc&start=0

Worth noting there are multiple versions of the SCUMM adlib format, v3 & v4 played correctly in adplay last time I tried. It's the later versions this bug is about. ScummVM in Adlib mode plays them flawlessly in Adlib mode, maybe their code is worth a look:
https://github.com/scummvm/scummvm/blob/master/audio/adlib.cpp

**Comment by mywave82** [comment]
Is this the correct music version?
https://youtu.be/bqOZmoF2u_o?t=115

**Comment by JPLeBreton** [comment]
> Is this the correct music version?

Yes, that sounds like correct Adlib playback. I'm assuming that video was captured from either DOSBox or ScummVM.

**Comment by JPLeBreton** [comment]
Looking at this again ~6 years later now that I know a bit more (though not enough to write any code for a PR, sadly) about Adlib and SCUMM:

LAA files are probably little more than MIDI data with a different header and maybe a few bytes of extra stuff. As far as I can tell, the LAA file extension was invented by the Adplug project to distinguish it from another existing totally unrelated (afaict) format called ADL. But internally, in the LucasArts game resource files that comprise the games, LAA data is called ADL (to differentiate it from ROL, ie MIDI data for the Roland sound card, and GMD, ie totally vanilla General MIDI data):
https://wiki.scummvm.org/index.php?title=SCUMM/Technical_Reference/Sound_resources
https://wiki.scummvm.org/index.php?title=SCUMM/Technical_Reference/SCUMM_6_resource_files#3.2.21_SOUN

I can't remember exactly when or where or how I got it, but I have an archive of LAAs that are pretty much complete rips from these LucasArts games (in order of release year / SCUMM version number): Indiana Jones and the Last Crusade, Secret of Monkey Island, Monkey Island 2, Indiana Jones and the Fate of Atlantis, and Day of the Tentacle. I'm happy to share this archive, it's a little over 1MB.

Listening to any of these LAAs from the command line with Adplay, the timing of the MIDI data sounds totally accurate. And the instrument _mappings_ seem more or less correct - if a note is a Piano in the original soundtrack, it generally sounds like a generic piano here. But the FM synth _instrument definitions_ sound way off.

So the fix for this issue might be to figure out where the FM instrument definition data is stored, and reading that and passing it along to the emulated OPL device. What I'm not sure of is where this data is stored - like I said, I don't think it's embedded within each LAA file; that would have added some redundant data to each file, and old games were very sparing about disk space usage - but I could be wrong.

Looking at the documentation for the iMUSE system, which LucasArts games from Monkey 2 onward, there is a bit at the bottom of this page about instrument data:
https://wiki.scummvm.org/index.php?title=SCUMM/Technical_Reference/iMuse_data
So it's possible that iMUSE games used that system to change instruments on the fly. And maybe they used it to set entire instrument banks on startup. Or maybe the initial banks are set elsewhere, from some data.
One thing I know is that ScummVM's Adlib emulation for these games sounds perfect, and ScummVM requires the original game resource files _but not the original game EXE_. So if the instrument data is stored in the executable, ScummVM has recreated it somewhere. Alternately, it's reading it from somewhere else in the game resource files. I suspect anyone who's worked even a bit with this part of the ScummVM code could tell us answers to these simple questions, which would likely illuminate a lot here.

**Comment by mywave82** [comment]
LAA files seems be "MIDI" dumps from SCUMM:
convertADResource() in https://github.com/scummvm/scummvm/blob/master/engines/scumm/sound.cpp

Running adplug like this:
adplay --stereo -e nuked -O disk -d myrecord.wav mi2_monkey_island_2_introduction.laa
and then comparing it with the audio from the youtube video does not reveal any huge differences except:
* intro has different length, LAA file is from a different revision of the game data, or SCUMM
* the general tempo is slightly off
* audio shape is off, but youtube audio compression is likely big part of the blame here

We will need to have a fresh clean LAA dump from SCUMM, and a clean .WAV rip from SCUMM that are from the same game-data and version of SCUMM if we are to make any improvements.


Adplug plays these files using mid.cpp

**Comment by Malvineous** [comment]
If it helps, I wrote the [dro2txt](https://github.com/Malvineous/libgamemusic/blob/master/utils/dro2txt.cpp) utility some years ago to help check the accuracy of OPL music.  This is simpler than comparing waveforms.

It works by dumping a list of changed registers each time a note is played.  The idea is to record an OPL dump as a .dro file in DOSBox, and then have your player write its OPL data to a .dro file as well.  This utility is used to convert both .dro files to .txt and then the `diff` command (or equivalent comparison tool) is used to check for any differences in the files.

Because the registers are only dumped as text when notes are playing, it can be used to compare songs even if the registers were written in a different order, which makes the files very different at the byte level.  As long as the OPL registers all contain the same values when a note is played, the .txt output will be identical.

Since AdPlug can only output .raw and not .dro files, you'd either have to add .dro writing support, or convert the .raw to .dro with libgamemusic first, so it won't be an out-of-the-box solution.  But it should make it much easier to see if there are any differences between the game's OPL output and AdPlug's, and help pinpoint which registers, if any, are being written incorrectly.

**Comment by moralrecordings** [comment]
The instruments are in the ADL block, stored as SysEx messages. I compared the OPL2 register output between ScummVM and AdPlug; it seems that AdPlug isn't sending correct values to the Level Key Scaling register, which affects e.g. how brassy the notes sound in the Monkey 2 intro. ScummVM uses a volume lookup table to calculate this value. 

ScummVM:
- read value from SysEx message: https://github.com/scummvm/scummvm/blob/master/audio/adlib.cpp#L1376
- determine volume: https://github.com/scummvm/scummvm/blob/master/audio/adlib.cpp#L2026
- write value to register: https://github.com/scummvm/scummvm/blob/master/audio/adlib.cpp#L2114

AdPlug:
- read value from SysEx message - https://github.com/adplug/adplug/blob/master/src/mid.cpp#L736
- write value to register: https://github.com/adplug/adplug/blob/master/src/mid.cpp#L393

**Comment by mywave82** [comment]
(you are not forgotten, but I am working with other items at the moment)

@moralrecordings To speed things up, you can attempt to write the change yourself and submit a PR. Seems like you already are deep-diving into this issue.
